transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -329,11 +332,35 @@ protected void run() {
                         break;
 
                     case SelectStrategy.SELECT:
-                        if (wakenUp == 1) {
-                            wakenUp = 0;
+                        if (pendingWakeup) {
+                            // We are going to be immediately woken so no need to reset wakenUp
+                            // or check for timerfd adjustment.
+                            strategy = epollWaitTimeboxed();
+                            if (strategy != 0) {
+                                break;
+                            }
+                            // We timed out so assume that we missed the write event due to an
+                            // abnormally failed syscall (the write itself or a prior epoll_wait)
+                            pendingWakeup = false;
+                            if (hasTasks()) {
+                                break;
+                            }
+                            // fall-through
                         }
-                        if (!hasTasks()) {
-                            strategy = epollWait();
+
+                        // Ordered store is sufficient here since the only access outside this
+                        // thread is a getAndSet in the wakeup() method
+                        wakenUp.lazySet(0);
```

Comment:
was lazySet safe here based on what @njhill said the other day?

Reply:
@johnou good question. To distil the problem, say we have fields
```java
final AtomicInteger x = new AtomicInteger(1), y = new AtomicInteger(0);
```
Thread 1 program order:
```java
x.lazySet(0);
if (y.get() == 0) { print("sleep"); }
```
Thread 2 program order:
```java
y.compareAndSet(0, 1); // (always succeeds)
if (x.compareAndSet(0, 1)) { print("wake"); }
```
Is it possible for "sleep" to be printed without "wake" also being printed? I now think the answer is yes, even more so after reading [this](https://groups.google.com/forum/#!msg/mechanical-sympathy/yeBNqThRtdU/weQY0Cy5PhcJ). Which means that `lazySet` _isn't_ safe, but would be good to get confirmation from others... @franz1981?? :)

Reply:
You've said right: it isn't safe and set would be a better choice for a total ordered operation (among the others) :)

Reply:
@franz1981 please review the rest as well